core/src/main/java/com/linecorp/armeria/internal/common/Http1ObjectEncoder.java

```diff
@@ -112,13 +119,48 @@ protected final ChannelFuture writeNonInformationalHeaders(int id, HttpObject co
         } else {
             f = write(id, converted, false);
             if (endStream) {
-                f = write(id, LastHttpContent.EMPTY_LAST_CONTENT, true);
+                final ChannelPromise promise = channel().newPromise();
+                allOf(promise, ImmutableList.of(f, write(id, LastHttpContent.EMPTY_LAST_CONTENT, true)));
+                f = promise;
```

Comment:
- How about implementing a special aggregator for two `Promise`?
- As I commented in https://github.com/line/armeria/issues/4096#issuecomment-1068684129, could we optionally aggregate the promises using a sampler?

Reply:
* You mean not list but just two ChannelFuture?
* https://github.com/line/armeria/pull/4149/files#diff-07e331c9d29de2371ad57fffd33330d8882b733c263081f5ed7aa583be3811c5R145 I did it here. Even is sample return false, should we return first exception or 2nd exception? Originally I'm thinking even false, we better to return 1st one without suppressed info.

Reply:
Like this?
```
private static void combine(ChannelPromise resultPromise, ChannelFuture first, ChannelFuture second) {
        final GenericFutureListener<Future<? super Void>> listener =
                new GenericFutureListener<Future<? super Void>>() {
                    private final AtomicInteger counter = new AtomicInteger(2);

                    @Override
                    public void operationComplete(Future<? super Void> ignore) throws Exception {
                        if (counter.decrementAndGet() == 0) {
                            final Throwable firstCause = first.cause();
                            final Throwable secondCause = second.cause();
                            if (firstCause != null) {
                                if (Flags.verboseExceptionSampler().isSampled(firstCause.getClass()) &&
                                    secondCause != null) {
                                    firstCause.addSuppressed(secondCause);
                                }
                                resultPromise.setFailure(firstCause);
                                return;
                            }
                            if (secondCause != null) {
                                resultPromise.setFailure(firstCause);
                                return;
                            }
                            resultPromise.setSuccess();
                        }
                    }
                };
        first.addListener(listener);
        second.addListener(listener);
    }
```

Reply:
I think @ikhoon was worried about the overhead of combining two futures into one, so I think it's not worth using the sampler inside the new promise.

Reply:
Does it mean we keep using 2nd Future's error if not sampled?

Reply:
I didn't express my intention well. I added a commit to apply a sampler before combining two promises.